logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/backup/RestoreBackupUseCase.kt

```
@@ -1,38 +1,121 @@
      */
-    suspend operator fun invoke(backupFilePath: String): RestoreBackupResult
+    suspend operator fun invoke(extractedBackupRootPath: Path, password: String?): RestoreBackupResult
 }
 
 internal class RestoreBackupUseCaseImpl(
-    private val backupImporter: BackupImporter,
-    private val dispatchers: KaliumDispatcher = KaliumDispatcherImpl
+    private val databaseImporter: DatabaseImporter,
+    private val kaliumFileSystem: KaliumFileSystem,
+    private val userId: UserId,
+    private val dispatchers: KaliumDispatcher = KaliumDispatcherImpl,
 ) : RestoreBackupUseCase {
 
-    override suspend operator fun invoke(backupFilePath: String): RestoreBackupResult {
-        return withContext(dispatchers.io) {
-            wrapStorageRequest {
-                backupImporter.importFromFile(backupFilePath)
-            }.fold({ RestoreBackupResult.Failure(it) }, { RestoreBackupResult.Success })
+    override suspend operator fun invoke(extractedBackupRootPath: Path, password: String?): RestoreBackupResult =
+        withContext(dispatchers.io) {
+            val idMapper = MapperProvider.idMapper()
+            val encryptedFilePath = kaliumFileSystem.list(extractedBackupRootPath).firstOrNull { it.name.contains(".cc20") }
+            val isPasswordProtected = !password.isNullOrEmpty()
+
+            // Check if the backup file is encrypted and if the password is provided
+            if (encryptedFilePath != null && !isPasswordProtected) return@withContext RestoreBackupResult.Failure(
+                DecryptionFailure("A non blank password is required to restore a password protected backup")
+            )
+            if (encryptedFilePath == null && isPasswordProtected) return@withContext RestoreBackupResult.Failure(
+                DecryptionFailure("A password was provided but no encrypted backup file was found")
+            )
+
+            if (isPasswordProtected) {
```

I am not sure I follow the logic here, if the encryptedFilepath is null do we even care about anything else ? we should return the error right ? 

I am also not sure why  isPasswordProtected is validated by looking if the user provided any password for the use-case

---

If the `encryptedFilepath` is null and there is no password, it means the backup was not encrypted and we just proceed with the restore. But if the `encryptedFilepath` is null and there is a password, then an unexpected error happened and we return it.

---

I refactored this logic a bit and put it in a separate method to improve slightly the readability.